src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -1205,13 +1232,75 @@ - (void)_handleEventReport:(NSArray<NSDictionary<NSString *, id> *> *)eventRepor
     }
 }
 
+#ifdef DEBUG
+- (void)unitTestClearClusterData
+{
+    std::lock_guard lock(_lock);
+    NSAssert(_persistedClusterData != nil, @"Test is not going to test what it thinks is testing!");
+    [_persistedClusterData removeAllObjects];
+}
+#endif
+
+- (nullable MTRDeviceClusterData *)_clusterDataForPath:(MTRClusterPath *)clusterPath
+{
+    os_unfair_lock_assert_owner(&self->_lock);
+
+    if (_clusterDataToPersist != nil) {
+        // Use the "dirty" values, if we have them.
+        MTRDeviceClusterData * data = _clusterDataToPersist[clusterPath];
+        if (data != nil) {
+            return data;
+        }
+    }
+
+    if (_persistedClusterData != nil) {
+        MTRDeviceClusterData * data = [_persistedClusterData objectForKey:clusterPath];
+        if (data != nil) {
+            return data;
+        }
+    }
+
+    if (![_persistedClusters containsObject:clusterPath]) {
+        // We are not expected to have this cluster, so no point in paging it in
+        // loading it from storage.
+        return nil;
+    }
+
+    NSAssert(_deviceController.controllerDataStore != nil,
+        @"How can _persistedClusters have an entry if we have no persistence?");
+    NSAssert(_persistedClusterData != nil,
+        @"How can _persistedClusterData not exist if we have persisted clusters?");
+
+    // Page in the stored value for the data.
+    MTRDeviceClusterData * data = [_deviceController.controllerDataStore getStoredClusterDataForNodeID:_nodeID endpointID:clusterPath.endpoint clusterID:clusterPath.cluster];
+    if (data != nil) {
+        [_persistedClusterData setObject:data forKey:clusterPath];
+    }
```

Comment:
If at this point data is nil (e.g. storage data loss, or previous failure to store), should we trigger some kind of recovery mechanism?

Reply:
In general, yes, we should.  I am not sure what form that can/should take, yet.  But we should probably have a followup to figure that out and do that.

That said, we should also trigger that mechanism if the data that comes back is not valid, etc, right?  So might be a bigger task than just "we got nothing".

Reply:
Yup I don't see this as blocking merge, and definite more of a follow up task.